src/lsm/cache_map.zig

```diff
@@ -0,0 +1,376 @@
+            self.scope_is_active = false;
+
+            // We don't need to do anything to persist a scope.
+            if (mode == .persist) {
+                self.scope_map.clearRetainingCapacity();
+                return;
+            }
+
+            // The scope_map stores the operations we need to reverse the changes a scope made.
+            // Replay them back.
+            var scope_values = self.scope_map.keyIterator();
+            while (scope_values.next()) |scope_value| {
+                if (tombstone(scope_value)) {
+                    // Reverting an insert consists of a .remove call. The value in here will be a
+                    // tombstone indicating the original value didn't exist. We don't touch map_2;
+                    // since we can never insert into it directly (only a .compact() can).
+                    const key = key_from_value(scope_value);
+                    _ = self.cache.remove(key);
+                    _ = self.map_1.remove(tombstone_from_key(key));
+                } else {
+                    // Reverting an update or delete consists of an insert of the original value.
+                    self.upsert(scope_value);
+                }
+            }
+            self.scope_map.clearRetainingCapacity();
+        }
+
+        pub fn compact(self: *Self) void {
+            assert(!self.scope_is_active);
+            assert(self.scope_map.count() == 0);
+
+            self.map_2.clearRetainingCapacity();
+            std.mem.swap(Map, &self.map_1, &self.map_2);
+        }
+    };
+}
+
+pub const TestTable = struct {
+    pub const Key = u32;
+    pub const Value = struct {
+        key: Key,
+        value: u32,
+        tombstone: bool = false,
+        padding: [7]u8 = undefined,
+    };
+
+    pub inline fn key_from_value(v: *const Value) u32 {
+        return v.key;
+    }
+
+    pub inline fn compare_keys(a: Key, b: Key) std.math.Order {
+        return std.math.order(a, b);
+    }
+
+    pub inline fn tombstone_from_key(a: Key) Value {
+        return Value{ .key = a, .value = 0, .tombstone = true };
+    }
+
+    pub inline fn tombstone(a: *const TestTable.Value) bool {
+        return a.tombstone;
+    }
+
+    pub inline fn hash(key: TestTable.Key) u64 {
+        return stdx.hash_inline(key);
+    }
+
+    pub inline fn equal(a: TestTable.Key, b: TestTable.Key) bool {
+        return a == b;
+    }
+};
+
+pub const TestCacheMap = CacheMapType(
+    TestTable.Key,
+    TestTable.Value,
+    TestTable.key_from_value,
+    TestTable.hash,
+    TestTable.equal,
+    TestTable.tombstone_from_key,
+    TestTable.tombstone,
+);
+
+test "cache_map: unit" {
+    const testing = std.testing;
+
+    const allocator = testing.allocator;
+
+    var cache_map = try TestCacheMap.init(allocator, .{
+        .cache_value_count_max = 2048,
+        .scope_value_count_max = 32,
+        .map_value_count_max = 32,
+        .name = "test map",
+    });
+    defer cache_map.deinit(allocator);
```

Comment:
Even for unit-tests, it might be a good idea to include comparison with the model. 

That is, fuzz tests check `model == implementation`, while the unit tests check

```
model == gold_value and implementation == gold_value
```

That way, you also test that the model behaves as you expect, and you actually share the infra for fuzz and unit-tests.

Reply:
Ah, clever, I like that!